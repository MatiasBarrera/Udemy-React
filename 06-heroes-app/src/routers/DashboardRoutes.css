.container {
  padding: 1rem 5rem;
  height: 90vh;
}
