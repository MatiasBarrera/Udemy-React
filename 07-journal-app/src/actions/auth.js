import {
  getAuth,
  signInWithPopup,
  createUserWithEmailAndPassword,
  updateProfile,
  signInWithEmailAndPassword,
  signOut
} from '@firebase/auth'
import Swal from 'sweetalert2'
import { googleAuthProvider } from '../firebase/firebase-config'
import { types } from '../types/types'
import { FinishLoading, StartLoading } from './ui'

export const startLoginEmailPassword = (email, password) => {
  return dispatch => {
    const auth = getAuth()

    dispatch(StartLoading())
    signInWithEmailAndPassword(auth, email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName))
        dispatch(FinishLoading())
      })
      .catch(e => {
        Swal.fire({
          icon: 'error',
          title: 'Error',
          text: 'There is no user record corresponding to this identifier. The user may have been deleted.'
        })
        dispatch(FinishLoading())
      })
  }
}

export const startRegisterWithEmailPasswordName = (name, email, password) => {
  return dispatch => {
    const auth = getAuth()
    createUserWithEmailAndPassword(auth, email, password)
      .then(async ({ user }) => {
        await updateProfile(user, { displayName: name })

        dispatch(login(user.uid, user.displayName))
      })
      .catch(e => {
        console.log(e)
      })
  }
}

export const startGoogleLogin = () => {
  return dispatch => {
    const auth = getAuth()
    signInWithPopup(auth, googleAuthProvider).then(({ user }) => {
      dispatch(login(user.uid, user.displayName))
    })
  }
}

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    name: displayName
  }
})

export const startLogout = () => {
  return async dispatch => {
    const auth = getAuth()
    await signOut(auth)
    dispatch(logout())
  }
}

export const logout = () => ({
  type: types.logout
})
