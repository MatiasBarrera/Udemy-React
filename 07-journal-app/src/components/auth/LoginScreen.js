import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth'
import googleLogo from '../../assets/google.logo.svg'
import { useForm } from '../../hooks/useForm'

export const LoginScreen = () => {
  const dispatch = useDispatch()
  const { loading } = useSelector(state => state.ui)

  const [formValues, handleInputChange] = useForm({
    email: 'matias@gmail.com',
    password: '123456'
  })

  const handleLogin = e => {
    e.preventDefault()
    dispatch(startLoginEmailPassword(email, password))
  }

  const handleLoginGoogle = () => {
    dispatch(startGoogleLogin())
  }

  const { email, password } = formValues
  return (
    <>
      <h3 className="login__title">Login</h3>
      <form className="auth__form" onSubmit={handleLogin}>
        <input
          type="text"
          placeholder="email"
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="password"
          name="password"
          autoComplete="off"
          value={password}
          onChange={handleInputChange}
        />
        <button className="pointer" type="submit" disabled={loading}>
          Ingresar
        </button>
        <hr />

        <div className="login__social-network">
          <p>Login with social network</p>
          <div
            className="google-btn pointer mt-1 mb-1"
            onClick={handleLoginGoogle}>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src={googleLogo}
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>

        <Link to="/auth/register">Create new account</Link>
      </form>
    </>
  )
}
