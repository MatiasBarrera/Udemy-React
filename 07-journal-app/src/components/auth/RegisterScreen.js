import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import isEmail from 'validator/lib/isEmail'
import { useDispatch, useSelector } from 'react-redux'
import { removeError, setError } from '../../actions/ui'
import { startRegisterWithEmailPasswordName } from '../../actions/auth'

export const RegisterScreen = () => {
  const dispatch = useDispatch()
  const { msgError } = useSelector(state => state.ui)

  const [formValues, handleInputChange] = useForm({
    name: 'matias',
    email: 'matias@gmail.com',
    password: '123456',
    password2: '123456'
  })

  const { name, email, password, password2 } = formValues

  const handleRegister = e => {
    e.preventDefault()

    if (isFormValid()) {
      dispatch(startRegisterWithEmailPasswordName(name, email, password))
    }
  }

  const isFormValid = () => {
    if (name.trim().length === 0) {
      dispatch(setError('Name is required'))
      return false
    }
    if (email.trim().length === 0 && !isEmail(email)) {
      dispatch(setError('Email is not valid'))
      return false
    }
    if (password !== password2 || password.length < 5) {
      dispatch(
        setError(
          'Password should be at least 6 characters and match each other '
        )
      )
      return false
    }
    dispatch(removeError())
    return true
  }

  return (
    <>
      <h3 className="login__title">Register</h3>
      <form className="auth__form" onSubmit={handleRegister}>
        {msgError && <div className="auth__alert-error">{msgError}</div>}
        <input
          type="text"
          placeholder="name"
          name="name"
          autoComplete="off"
          value={name}
          onChange={handleInputChange}
        />
        <input
          type="text"
          placeholder="email"
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="password"
          name="password"
          autoComplete="off"
          value={password}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="confirm"
          name="password2"
          autoComplete="off"
          value={password2}
          onChange={handleInputChange}
        />
        <button className="pointer" type="submit">
          Register
        </button>
        <hr />

        <Link to="/auth/login">Already registered?</Link>
      </form>
    </>
  )
}
