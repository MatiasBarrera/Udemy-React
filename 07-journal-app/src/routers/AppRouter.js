import { getAuth, onAuthStateChanged } from '@firebase/auth'
import React, { useEffect, useState } from 'react'
import { useDispatch } from 'react-redux'
import { login } from '../actions/auth'

import {
  BrowserRouter as Router,
  Switch,
  // Route,
  Redirect
} from 'react-router-dom'
import { JournalScreen } from '../components/journal/JournalScreen'
import { AuthRouter } from './AuthRouter'
import { PublicRoute } from './PublicRoute'
import { PrivateRoute } from './PrivateRoute'

export const AppRouter = () => {
  const dispatch = useDispatch()

  const [checking, setChecking] = useState(true)

  const [isLoggedIn, setIsLoggedIn] = useState(false)

  useEffect(() => {
    const auth = getAuth()
    onAuthStateChanged(auth, user => {
      if (user?.uid) {
        dispatch(login(user.uid, user.displayName))
        setIsLoggedIn(true)
      } else {
        setIsLoggedIn(false)
      }
      setChecking(false)
    })
  }, [dispatch, setChecking, setIsLoggedIn, isLoggedIn])

  if (checking) {
    return <h1>Cargando...</h1>
  }

  return (
    <Router>
      <div>
        <Switch>
          <PublicRoute
            path="/auth"
            component={AuthRouter}
            isLoggedIn={isLoggedIn}
          />
          <PrivateRoute
            exact
            isLoggedIn={isLoggedIn}
            path="/"
            component={JournalScreen}
          />

          <Redirect to="/auth/login" />
        </Switch>
      </div>
    </Router>
  )
}
