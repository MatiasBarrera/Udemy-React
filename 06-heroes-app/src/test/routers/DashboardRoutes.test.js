import React from 'react'
import { mount } from 'enzyme'
import { DashboardRoutes } from '../../routers/DashboardRoutes'
import { MemoryRouter } from 'react-router'
import { AuthContext } from '../../auth/AuthContext'

describe('Pruebas en DashboardRoutes', () => {
  const contextValue = {
    dispatch: jest.fn(),
    user: {
      logged: true,
      name: 'Matias'
    }
  }

  test('Debe de mostrar el componente correctamente ', () => {
    const wrapper = mount(
      <MemoryRouter>
        <AuthContext.Provider value={contextValue}>
          <DashboardRoutes></DashboardRoutes>
        </AuthContext.Provider>
      </MemoryRouter>
    )

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.container-user').text().trim()).toBe('Matias')
  })
})
