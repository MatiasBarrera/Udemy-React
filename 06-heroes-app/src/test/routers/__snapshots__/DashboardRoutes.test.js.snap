// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en DashboardRoutes Debe de mostrar el componente correctamente  1`] = `
Array [
  <nav>
    <a
      className="brand"
      href="/"
      onClick={[Function]}
    >
      Asociaciones
    </a>
    <div
      className="container-links"
    >
      <a
        aria-current="page"
        className="active"
        href="/marvel"
        onClick={[Function]}
        style={Object {}}
      >
        Marvel
      </a>
      <a
        aria-current={null}
        href="/dc"
        onClick={[Function]}
      >
        DC Comics
      </a>
      <a
        aria-current={null}
        href="/search"
        onClick={[Function]}
      >
        Search
      </a>
    </div>
    <div
      className="container-user"
    >
      Matias
    </div>
    <div
      className="container-logout"
    >
      <button
        onClick={[Function]}
      >
        Logout
      </button>
    </div>
  </nav>,
  <div
    className="container"
  >
    <div>
      <h1>
        Marvel Screen
      </h1>
      <br />
      <div
        className="card-columns"
      >
        <div
          className="card animate__fadeIn animate__animated animate__fast"
        >
          <div>
            <img
              alt="Spider Man"
              className="card-img"
              src=""
            />
          </div>
          <div
            className="card-body"
          >
            <h5>
              Spider Man
            </h5>
            <p
              className="card-text"
            >
              Peter Parker
            </p>
            <p
              className="card-text"
            >
              <small
                className="text-muted"
              >
                Amazing Fantasy #15
              </small>
            </p>
            <a
              className="link"
              href="/hero/marvel-spider"
              onClick={[Function]}
            >
              Mas...
            </a>
          </div>
        </div>
        <div
          className="card animate__fadeIn animate__animated animate__fast"
        >
          <div>
            <img
              alt="Captain America"
              className="card-img"
              src=""
            />
          </div>
          <div
            className="card-body"
          >
            <h5>
              Captain America
            </h5>
            <p
              className="card-text"
            >
              Steve Rogers
            </p>
            <p
              className="card-text"
            >
              <small
                className="text-muted"
              >
                Captain America Comics #1
              </small>
            </p>
            <a
              className="link"
              href="/hero/marvel-captain"
              onClick={[Function]}
            >
              Mas...
            </a>
          </div>
        </div>
        <div
          className="card animate__fadeIn animate__animated animate__fast"
        >
          <div>
            <img
              alt="Iron Man"
              className="card-img"
              src=""
            />
          </div>
          <div
            className="card-body"
          >
            <h5>
              Iron Man
            </h5>
            <p
              className="card-text"
            >
              Tony Stark
            </p>
            <p
              className="card-text"
            >
              <small
                className="text-muted"
              >
                Tales of Suspense #39
              </small>
            </p>
            <a
              className="link"
              href="/hero/marvel-iron"
              onClick={[Function]}
            >
              Mas...
            </a>
          </div>
        </div>
        <div
          className="card animate__fadeIn animate__animated animate__fast"
        >
          <div>
            <img
              alt="Thor"
              className="card-img"
              src=""
            />
          </div>
          <div
            className="card-body"
          >
            <h5>
              Thor
            </h5>
            <p
              className="card-text"
            >
              Thor Odinson
            </p>
            <p
              className="card-text"
            >
              <small
                className="text-muted"
              >
                Journey into Myster #83
              </small>
            </p>
            <a
              className="link"
              href="/hero/marvel-thor"
              onClick={[Function]}
            >
              Mas...
            </a>
          </div>
        </div>
        <div
          className="card animate__fadeIn animate__animated animate__fast"
        >
          <div>
            <img
              alt="Hulk"
              className="card-img"
              src=""
            />
          </div>
          <div
            className="card-body"
          >
            <h5>
              Hulk
            </h5>
            <p
              className="card-text"
            >
              Bruce Banner
            </p>
            <p
              className="card-text"
            >
              <small
                className="text-muted"
              >
                The Incredible Hulk #1
              </small>
            </p>
            <a
              className="link"
              href="/hero/marvel-hulk"
              onClick={[Function]}
            >
              Mas...
            </a>
          </div>
        </div>
        <div
          className="card animate__fadeIn animate__animated animate__fast"
        >
          <div>
            <img
              alt="Wolverine"
              className="card-img"
              src=""
            />
          </div>
          <div
            className="card-body"
          >
            <h5>
              Wolverine
            </h5>
            <p
              className="card-text"
            >
              James Howlett
            </p>
            <p
              className="card-text"
            >
              <small
                className="text-muted"
              >
                The Incredible Hulk #180
              </small>
            </p>
            <a
              className="link"
              href="/hero/marvel-wolverine"
              onClick={[Function]}
            >
              Mas...
            </a>
          </div>
        </div>
        <div
          className="card animate__fadeIn animate__animated animate__fast"
        >
          <div>
            <img
              alt="Daredevil"
              className="card-img"
              src=""
            />
          </div>
          <div
            className="card-body"
          >
            <h5>
              Daredevil
            </h5>
            <p
              className="card-text"
            >
              Matthew Michael Murdock
            </p>
            <p
              className="card-text"
            >
              <small
                className="text-muted"
              >
                Daredevil #1
              </small>
            </p>
            <a
              className="link"
              href="/hero/marvel-daredevil"
              onClick={[Function]}
            >
              Mas...
            </a>
          </div>
        </div>
        <div
          className="card animate__fadeIn animate__animated animate__fast"
        >
          <div>
            <img
              alt="Hawkeye"
              className="card-img"
              src=""
            />
          </div>
          <div
            className="card-body"
          >
            <h5>
              Hawkeye
            </h5>
            <p
              className="card-text"
            >
              Clinton Francis Barton
            </p>
            <p
              className="card-text"
            >
              <small
                className="text-muted"
              >
                Tales of Suspense #57
              </small>
            </p>
            <a
              className="link"
              href="/hero/marvel-hawkeye"
              onClick={[Function]}
            >
              Mas...
            </a>
          </div>
        </div>
        <div
          className="card animate__fadeIn animate__animated animate__fast"
        >
          <div>
            <img
              alt="Cyclops"
              className="card-img"
              src=""
            />
          </div>
          <div
            className="card-body"
          >
            <h5>
              Cyclops
            </h5>
            <p
              className="card-text"
            >
              Scott Summers
            </p>
            <p
              className="card-text"
            >
              <small
                className="text-muted"
              >
                X-Men #1
              </small>
            </p>
            <a
              className="link"
              href="/hero/marvel-cyclops"
              onClick={[Function]}
            >
              Mas...
            </a>
          </div>
        </div>
        <div
          className="card animate__fadeIn animate__animated animate__fast"
        >
          <div>
            <img
              alt="Silver Surfer"
              className="card-img"
              src=""
            />
          </div>
          <div
            className="card-body"
          >
            <h5>
              Silver Surfer
            </h5>
            <p
              className="card-text"
            >
              Norrin Radd
            </p>
            <p
              className="card-text"
            >
              <small
                className="text-muted"
              >
                The Fantastic Four #48
              </small>
            </p>
            <a
              className="link"
              href="/hero/marvel-silver"
              onClick={[Function]}
            >
              Mas...
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
